common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+    /// ever be allowed through, as it's bigger than the size of the bucket.
+    pub fn acquire_all_tokens(&mut self, requested: usize) -> Result<(), Option<Instant>> {
+        // Skip over if we purposely have an open throttle
+        if self.open || requested == 0 {
+            return Ok(());
+        }
+
+        // Refill if needed
+        self.refill();
+
+        if self.tokens >= requested {
+            self.deduct_tokens(requested);
+            Ok(())
+        } else {
+            Err(self.time_of_tokens_needed(requested))
+        }
+    }
+
+    /// Returns `usize` of tokens allowed.  May be less than requested.
+    /// For best effort, caller should return unused tokens with `add_tokens`
+    pub fn acquire_tokens(&mut self, requested: usize) -> Result<usize, Instant> {
+        // Skip over if we purposely have an open throttle
+        if self.open || requested == 0 {
+            return Ok(requested);
+        }
+
+        // Refill if needed
+        self.refill();
+
+        let allowed = self.deduct_tokens(requested);
+        if allowed > 0 {
+            Ok(allowed)
+        } else {
+            Err(self.time_of_next_refill())
+        }
+    }
+
+    /// Retrieve the maximum amount of tokens up to `count`
+    /// Tells us how much of the requested size we can send
+    fn deduct_tokens(&mut self, requested: usize) -> usize {
+        let tokens = self.tokens;
+        let tokens_left = tokens.saturating_sub(requested);
+        let tokens_allowed = min(tokens, requested);
+        self.tokens = tokens_left;
+
+        tokens_allowed
+    }
+
+    /// Tells us when the next refill is
+    pub fn time_of_next_refill(&self) -> Instant {
+        self.last_refresh_time + ONE_SEC
+    }
+
+    /// Tells us when an entire batch will make it through.  Useful for Async work to wait until
+    /// all tokens are ready.  Returns `None` if it is never possible.
+    pub fn time_of_tokens_needed(&self, requested: usize) -> Option<Instant> {
+        if self.open {
+            Some(Instant::now())
+        } else if self.size.get() < requested {
+            // This means the batch can never succeed
+            None
+        } else {
+            let tokens_needed = requested.saturating_sub(self.tokens);
+
+            let intervals = (tokens_needed as f64 / self.rate.get() as f64).ceil() as u32;
+            Some(self.last_refresh_time + (ONE_SEC * intervals))
+        }
+    }
+
+    /// Add new tokens, this can also be used to add unused tokens back
+    /// Ensures bucket doesn't overfill
+    pub fn add_tokens(&mut self, new_tokens: usize) {
+        self.tokens = min(self.size.get(), self.tokens.saturating_add(new_tokens));
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{sync::MutexGuard, thread::sleep};
+    use tokio::time::Duration;
+
+    // Helper function, checks the throttle limit
+    fn assert_num_allowed(rate_limiter: &mut MutexGuard<Bucket>, num_allowed: usize) {
```

Comment:
feels like this should be called assert_acquire

Reply:
also why not just pass in the arc bucket instead of the mutex guard?

Reply:
> also why not just pass in the arc bucket instead of the mutex guard?

I think I was trying to avoid having to write `.lock()` everywhere